engines/dragons/cutscene.cpp

```diff
@@ -0,0 +1,888 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+#include "dragons/cutscene.h"
+#include "dragons/dragons.h"
+#include "dragons/actor.h"
+#include "dragons/actorresource.h"
+#include "dragons/background.h"
+#include "dragons/inventory.h"
+#include "dragons/cursor.h"
+#include "dragons/dragonini.h"
+#include "dragons/scene.h"
+#include "dragons/screen.h"
+#include "dragons/talk.h"
+
+namespace Dragons {
+
+CutScene::CutScene(DragonsEngine *vm): _vm(vm) {
+	loadPalettes();
+}
+
+CutScene::~CutScene() {
+	if (_palettes) {
+		free(_palettes);
+	}
+}
+
+void CutScene::scene1() {
+	// TODO spcLoadScene1 knights around the table.
+	bool isFlag10Set = _vm->isFlagSet(ENGINE_FLAG_10);
+
+	DragonINI *flicker = _vm->_dragonINIResource->getFlickerRecord();
+
+	DAT_80063514 = 0xb00;
+	_vm->_dragonINIResource->setFlickerRecord(NULL);
+
+	_vm->setUnkFlags(ENGINE_UNK1_FLAG_2);
+	//fade_related_calls_with_1f();
+	_vm->clearFlags(ENGINE_FLAG_10);
+
+	_vm->_inventory->setActorFlag400();
+	_vm->_cursor->setActorFlag400();
+
+//	scr_tilemap1_w = 0x28;
+//	DAT_8006a3f0 = DAT_8006a3ec;
+
+//	load_actor_file(0x81);
+//	load_actor_file(0x7d);
+//	load_actor_file(0x7e);
+//	load_actor_file(0x8f);
+//	load_actor_file(0xaa);
+	_vm->setFlags(ENGINE_FLAG_20000);
+	wideAngleEveryoneAtTable();
+	_vm->waitForFrames(0x5a);
+	closeUpShotOnActor(0xd3, 0, 0x233, 0x17a); //close up on king
+	//playSoundFromTxtIndex(0x4e26);
+	if (_vm->_talk->somethingTextAndSpeechAndAnimRelated(DAT_80072de8,4,0,0x4e26,0x2e01) != 2) {
+		fadeScreenAndResetActor(DAT_80072de8);
+		closeUpShotOnActor(0xd8, 0, 0xfd, 0x60); // cut to flicker
+		//playSoundFromTxtIndex(0x4ea2);
+		if (_vm->_talk->somethingTextAndSpeechAndAnimRelated(DAT_80072de8,2,0,0x4ea2,0x701) != 2) {
```

Comment:
Code complexity of the whole thing is tremendous.

I recommend to rewrite it like this:
```
struct {
   int dat;
   int par1, par2, par3, par3;
} const talkParams[] = {
   { DAT_80072de8, 5, 0, 0x4eec, 0x2e01 },
   { DAT_80072de8, 2, 0, 0x5152, 0x3e01 },
   ....
   { 0, 0, 0, 0, 0 }
};

for (i = 0; talkParams[i].dat; i++) {
   if (!_vm->_talk->somethingTextAndSpeechAndAnimRelated(talkParams[i].dat, talkParams[i].par1, ..))
      break;

   switch (i) {
   case 0:
     fadeScreenAndResetActor(DAT_80072de8);
     wideAngleEveryoneAtTable();  // shot of whole room
     break;
  case 1:
     _vm->_talk->somethingTextAndSpeechAndAnimRelated(DAT_800830c0,0x1d,0x1c,0x5074,0x501) != 2 &&
     _vm->_talk->somethingTextAndSpeechAndAnimRelated(DAT_80072df0,9,5,0x511c,0xc01) != 2 &&
     _vm->_talk->somethingTextAndSpeechAndAnimRelated(DAT_800830c0,0x1d,0x1c,0x5138,0x501) != 2) {
     closeUpShotOnActor(0xd7, 0, 0x312, 0x260); //close up on flame
     //playSoundFromTxtIndex(0x5152);
    break;
    .....
    default:
      break;
   }
}
```